raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java

```diff
@@ -0,0 +1,215 @@
+    private final MetricName numVoterConnectionsMetricName;
+    private final Sensor commitTimeSensor;
+    private final Sensor electionTimeSensor;
+    private final Sensor fetchRecordsSensor;
+    private final Sensor appendRecordsSensor;
+    private final Sensor pollIdleSensor;
+
+    public KafkaRaftMetrics(Metrics metrics, String metricGrpPrefix, QuorumState state, long timestamp) {
+        this.metrics = metrics;
+        this.bootTimestamp = timestamp;
+        String metricGroupName = metricGrpPrefix + "-metrics";
+
+        this.pollStartMs = 0L;
+        this.pollEndMs = 0L;
+        this.electionStartMs = -1L;
+        this.numVoterConnections = 0;
+        this.logEndOffset = new OffsetAndEpoch(0L, 0);
+
+        this.currentStateMetricName = metrics.metricName("current-state", metricGroupName, "The current state of this member; possible values are leader, candidate, voter, observer");
+        Gauge<String> stateProvider = (mConfig, currentTimeMs) -> {
+            if (state.isLeader()) {
+                return "leader";
+            } else if (state.isCandidate()) {
+                return "candidate";
+            } else {
+                if (state.isVoter())
+                    return "voter";
+                else
+                    return "observer";
+            }
+        };
+        metrics.addMetric(this.currentStateMetricName, null, stateProvider);
+
+        this.currentLeaderIdMetricName = metrics.metricName("current-leader", metricGroupName, "The current quorum leader's id; -1 indicates unknown");
+        metrics.addMetric(this.currentLeaderIdMetricName, (mConfig, currentTimeMs) -> state.leaderId().orElse(-1));
+
+        this.currentVotedIdMetricName = metrics.metricName("current-vote", metricGroupName, "The current voted leader's id; -1 indicates not voted for anyone");
+        metrics.addMetric(this.currentVotedIdMetricName, (mConfig, currentTimeMs) -> {
+            if (state.isLeader() || state.isCandidate()) {
+                return state.localId;
+            } else {
+                FollowerState followerState = state.followerStateOrThrow();
+                if (followerState.hasVoted())
+                    return followerState.votedId();
+                else
+                    return -1;
+            }
+        });
+
+        this.currentEpochMetricName = metrics.metricName("current-epoch", metricGroupName, "The current quorum epoch.");
+        metrics.addMetric(this.currentEpochMetricName, (mConfig, currentTimeMs) -> state.epoch());
+
+        this.highWatermarkMetricName = metrics.metricName("high-watermark", metricGroupName, "The high watermark maintained on this member; -1 if it is unknown");
+        metrics.addMetric(this.highWatermarkMetricName, (mConfig, currentTimeMs) -> state.highWatermark().orElse(-1L));
+
+        this.logEndOffsetMetricName = metrics.metricName("log-end-offset", metricGroupName, "The current raft log end offset.");
```

Comment:
Does the `ReplicatedLog#endOffset` incur additional cost than memory lookup? I checked `Log.scala` and it returns from memory IIUC:
```  
def logEndOffset: Long = nextOffsetMetadata.messageOffset
```
wondering if we could simply this metric by doing a lambda by passing in the `ReplicatedLog#endOffset`

Reply:
We can, but I'd prefer to not introduce the `log` into this metrics class.